server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Rights.scala

```diff
@@ -0,0 +1,162 @@
+    .toMap)
+}
+
+case class MailboxACL(entries: Map[EntryKey, Rfc4314Rights]) {
+  val asJava: JavaMailboxACL = {
+    val map: Map[EntryKey, JavaRfc4314Rights] = entries.view
+    .mapValues(seq => seq.asJava)
+    .toMap
+
+    new JavaMailboxACL(map.asJava)
+  }
+}
+
+object Right {
+  val UNSUPPORTED: Option[Boolean] = None
+
+  val Administer = Right(JavaRight.Administer)
+  val Expunge = Right(JavaRight.PerformExpunge)
+  val Insert = Right(JavaRight.Insert)
+  val Lookup = Right(JavaRight.Lookup)
+  val Read = Right(JavaRight.Read)
+  val Seen = Right(JavaRight.WriteSeenFlag)
+  val DeleteMessages = Right(JavaRight.DeleteMessages)
+  val Write = Right(JavaRight.Write)
+
+  private val allRights = Seq(Administer, Expunge, Insert, Lookup, Read, Seen, DeleteMessages, Write)
+
+  def forRight(right: JavaRight): Option[Right] = allRights.find(_.right.equals(right))
+
+  def forChar(c: Char): Option[Right] = allRights.find(_.asCharacter == c)
+}
+
+sealed case class Right(right: JavaRight) {
+  val asCharacter: Char = right.asCharacter
+
+  val toMailboxRight: JavaRight = right
+}
+
+object Rights {
+  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[Rights])
+
+  val EMPTY = new Rights(Map())
+
+  def of(username: Username, right: Right): Rights = of(username, Seq(right))
+
+  def of(username: Username, rights: Seq[Right]): Rights = {
+    require(rights.nonEmpty, "'rights' should not be empty")
+
+    Rights(Map(username -> rights))
+  }
+
+  def fromACL(acl: MailboxACL): Rights = acl.entries
+    .filter {
+      case (entryKey, _) => isSupported(entryKey)
+    }
+    .map {
+      case (entryKey, aclRights) => toRights(entryKey, aclRights)
+    }
+    .fold(EMPTY)(_ combine _)
+
+  private def toRights(entryKey: EntryKey, aclRights: Rfc4314Rights): Rights = of(Username.of(entryKey.getName), aclRights.toRights)
+
+  private def isSupported(key: EntryKey): Boolean = {
+    if (key.isNegative) {
+      LOGGER.info("Negative keys are not supported")
+      return false
+    }
+    if (key == JavaMailboxACL.OWNER_KEY) {
+      return false
+    }
+    if (key.getNameType ne JavaMailboxACL.NameType.user) {
+      LOGGER.info("{} is not supported. Only 'user' is.", key.getNameType)
+      return false
+    }
+    true
+  }
+}
+
+case class Rights private(rights: Map[Username, Seq[Right]]) {
+
+  def removeEntriesFor(username: Username) = Rights(rights.filter(!_._1.equals(username)))
+
+  def toMailboxAcl: MailboxACL = {
+    val map: Map[EntryKey, Rfc4314Rights] = rights.view
+      .mapValues(Rfc4314Rights.fromRights)
+      .toMap
+      .map {
+        case (user, rfc4314Rights) => (EntryKey.createUserEntryKey(user), rfc4314Rights)
+      }
+    MailboxACL(map)
+  }
+
+  def append(username: Username, right: Right): Rights = append(username, Seq(right))
+
+  def append(username: Username, rights: Seq[Right]): Rights = {
+    require(rights.nonEmpty, "'rights' should not be empty")
+
+    Rights(this.rights + (username -> rights))
```

Comment:
we usually use `copy` for that: as you will add properties in Rights, you won't have to come back to this copy code